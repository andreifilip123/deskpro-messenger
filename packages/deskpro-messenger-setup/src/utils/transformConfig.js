export default (settings) => {
  const screens = {
    index: {
      screenType: 'Blocks',
      blocks: []
    }
  };

  if (settings.getIn(['chat', 'enabled'])) {
    const chatBlockConfig = settings.getIn(['messenger', 'chat']);
    screens.index.blocks.push({
      blockType: 'StartChatBlock',
      title: chatBlockConfig.get('title', 'Conversation'),
      description: chatBlockConfig.get('description'),
      linkText: chatBlockConfig.get('buttonText', 'Start a chat'),
      showAgentPhotos: chatBlockConfig.get('showAgentPhotos'),
      to: 'startChat'
    });
    screens.startChat = settings.get('chat').toJS();
    screens.startChat.screenType = 'StartChatScreen';
    delete screens.startChat.enabled;
    if (screens.startChat.noAnswerBehavior === 'save_ticket') {
      screens.startChat.ticketFormConfig = [
        {
          fields: [
            {
              name: 'name',
              label: 'Full Name',
              type: 'text',
              validation: ['required'],
              placeholder: 'John Doe'
            },
            {
              name: 'email',
              label: 'Email',
              type: 'text',
              validation: ['required'],
              placeholder: 'john.doe@company.com'
            }
          ]
        }
      ];
    }
  }

  if (settings.getIn(['tickets', 'enabled'])) {
    const ticketBlockConfig = settings.getIn(['messenger', 'tickets']);
    screens.index.blocks.push({
      blockType: 'ScreenLink',
      blockTitle: ticketBlockConfig.get('title'),
      description: ticketBlockConfig.get('description'),
      label: ticketBlockConfig.get('buttonText'),
      to: 'newTicket'
    });
    screens.newTicket = settings.get('tickets').toJS();
    delete screens.newTicket.enabled;
  }

  const enabledGreetings = [null];
  if ('object' === typeof screens.startChat) {
    enabledGreetings.push('/screens/startChat');
  }
  const config = {
    screens,
    themeVars: {
      '--color-primary': settings.getIn(['styles', 'primaryColor'], '#3d88f3'),
      '--color-background': settings.getIn(['styles', 'backgroundColor'], '#3d88f3'),
      '--brand-primary': settings.getIn(['styles', 'primaryColor'], '#3d88f3'),
      '--brand-secondary': settings.getIn(['styles', 'backgroundColor'], '#a9b0b0'),
      '--text-color': settings.getIn(['styles', 'textColor'], '#ffffff'),

    },
    greetings: {},
    enabledGreetings
  };

  return config;
};
