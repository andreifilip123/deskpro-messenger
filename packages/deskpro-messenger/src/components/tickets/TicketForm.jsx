import React from 'react';
import { compose } from 'redux';
import { FieldLayout, Form, withFormik, TicketForm } from '@deskpro/portal-components';
import { FormattedMessage } from 'react-intl';

import Button from '../form/Button';
import withTranslatedLayout from '../core/TranslatedLayoutHOC';

const TicketForm = (props) => (
  <Form>
    <FieldLayout {...props} />
    <Button width="full" size="medium" color="primary" type="submit">
      <FormattedMessage
        id="tickets.form.submit_button"
        defaultMessage="Save Ticket"
      />
    </Button>
  </Form>
);

const formEnhancer = withFormik({
  enableReinitialize: true,
  mapPropsToValues: ({ layouts, initialValues = {} }) => {
    const layout = layouts.getMatchingLayout(initialValues);
    if (layout) {
      return { ...initialValues, ...layout.getDefaultValues() };
    }
    return { ...initialValues };
  },
  handleSubmit: (values, { props, setSubmitting }) => {
    // setSubmitting(true);
    props.onSubmit(values);
  }
});

export default compose(
  withTranslatedLayout,
  formEnhancer
)(TicketForm);
