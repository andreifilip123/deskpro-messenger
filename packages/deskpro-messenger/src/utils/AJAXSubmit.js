// eslint-disable-next-line func-names
const AJAXSubmit = (function () {
  function submitData(oData, config) {
    /* the AJAX request... */
    const oAjaxReq = new XMLHttpRequest();
    oAjaxReq.submittedData = oData;

    if (config.updateProgress) {
      oAjaxReq.addEventListener('progress', config.updateProgress);
    }
    if (config.transferComplete) {
      oAjaxReq.addEventListener('load', config.transferComplete);
    }
    if (config.transferFailed) {
      oAjaxReq.addEventListener('error', config.transferFailed);
    }
    if (config.transferCanceled) {
      oAjaxReq.addEventListener('abort', config.transferCanceled);
    }

    oAjaxReq.onreadystatechange = function() { // Call a function when the state changes.
      if (this.readyState === XMLHttpRequest.DONE) {
        if (this.status >= 400) {
          let message = this.statusText;
          if (this.response && this.response.errors && this.response.errors.file) {
            message = this.response.errors.file;
          }
          config.transferFailed({
            code: this.status,
            message
          });
        }
      }
    }

    oAjaxReq.withCredentials = true;
    /* method is POST */
    oAjaxReq.responseType = 'json';
    oAjaxReq.open('post', oData.receiver, true);
    /* enctype is multipart/form-data */
    const sBoundary = `---------------------------${Date.now().toString(16)}`;
    oAjaxReq.setRequestHeader('Content-Type', `multipart/form-data; boundary=${sBoundary}`);
    if (config.requestHeaders) {
      for (const [name, value] of Object.entries(config.requestHeaders)) {
        oAjaxReq.setRequestHeader(name, value);
      }
    }
    oAjaxReq.send(`--${sBoundary}\r\n${
      oData.segments.join(`--${sBoundary}\r\n`)}--${sBoundary}--\r\n`);
  }

  function processStatus(oData, config = {}) {
    if (oData.status > 0) { return; }
    /* the form is now totally serialized! do something before sending it to the server... */
    // /* doSomething(oData); */
    // /* console.log("AJAXSubmit - The form is now serialized. Submitting..."); */
    submitData(oData, config);
  }

  function pushSegment(oFREvt) {
    this.owner.segments[this.segmentIdx] += `${oFREvt.target.result}\r\n`;
    this.owner.status = this.owner.status - 1;
    processStatus(this.owner, this.config);
  }

  function SubmitRequest(config) {
    this.receiver = config.url;
    this.status = 0;
    this.segments = [];
    for (let nFile = 0; nFile < config.files.length; nFile++) {
      const oFile = config.files[nFile];
      const oSegmReq = new FileReader();
      /* (custom properties:) */
      oSegmReq.segmentIdx = this.segments.length;
      oSegmReq.owner = this;
      oSegmReq.config = config;
      /* (end of custom properties) */
      oSegmReq.onload = pushSegment;
      this.segments.push(`Content-Disposition: form-data; name="file"; filename="${oFile.name}"\r\nContent-Type: ${oFile.type}\r\n\r\n`);
      this.status = this.status + 1;
      oSegmReq.readAsBinaryString(oFile);
    }
    this.segments.push(`Content-Disposition: form-data;`);
    processStatus(this, config);
  }

  // eslint-disable-next-line func-names
  return function (config) {
    if (!config.url) { return; }
    new SubmitRequest(config);
  };
}());

export default AJAXSubmit;
