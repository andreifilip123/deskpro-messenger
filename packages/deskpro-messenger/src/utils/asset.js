window.parent.DESKPRO_MESSENGER_OPTIONS =
  window.parent.DESKPRO_MESSENGER_OPTIONS || {};

export default (path) => {
  let edited = '';
  if (window.parent.DESKPRO_MESSENGER_OPTIONS.baseUrl) {
    edited = window.parent.DESKPRO_MESSENGER_OPTIONS.baseUrl.replace(/\/$/, "");
  }
  return `${edited}/assets/${path}`;
};
