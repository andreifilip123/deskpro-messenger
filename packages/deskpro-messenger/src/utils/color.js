/**
 * Returns true if opposite color is light or not.
 * @param {string} color Color hash.
 * @link https://stackoverflow.com/a/35970186
 */
export const isLightColor = (color) => {
  let hex = color.startsWith('#') ? color.slice(1) : color;
  // convert 3-digit hex to 6-digits.
  if (hex.length === 3) {
    hex = hex[0] + hex[0] + hex[1] + hex[1] + hex[2] + hex[2];
  }
  if (hex.length !== 6) {
    throw new Error('Invalid HEX color.');
  }
  const r = parseInt(hex.slice(0, 2), 16),
    g = parseInt(hex.slice(2, 4), 16),
    b = parseInt(hex.slice(4, 6), 16);
  // http://stackoverflow.com/a/3943023/112731
  return r * 0.299 + g * 0.587 + b * 0.114 > 186;
};
