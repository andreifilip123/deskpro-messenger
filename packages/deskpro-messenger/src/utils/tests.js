import React from 'react';
import { render } from 'react-testing-library';
import { IntlProvider } from 'react-intl';
import { MemoryRouter } from 'react-router-dom';

const customRender = (node, { translations = {}, ...options } = {}) => {
  return render(
    <MemoryRouter>
      <IntlProvider locale="en-US" messages={translations}>
        {node}
      </IntlProvider>
    </MemoryRouter>,
    options
  );
};

// re-export everything
export * from 'react-testing-library';

// override render method
export { customRender as render };
