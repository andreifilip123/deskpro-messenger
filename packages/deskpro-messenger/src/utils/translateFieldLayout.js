/**
 * Translates all strings (labels) in a field layout config.
 *
 * @param {Array} config Field Layout config
 * @param {function} formatMessage Callback from react-intl to translate a string.
 *
 * @returns new translated field layout config.
 */
export default (config, formatMessage) => {
  const t = (str) => formatMessage({ id: str, defaultMessage: str });

  return config.map((layout) => ({
    ...layout,
    fields: layout.fields.map((field) => {
      const newField = { ...field };
      newField.label = t(field.label);
      if (newField.placeholder) {
        newField.placeholder = t(field.placeholder);
      }
      if (Array.isArray(newField.validation)) {
        newField.validation = field.validation.map(
          (rule) =>
            typeof rule === 'object'
              ? { ...rule, message: t(rule.message) }
              : rule
        );
      }
      if (Array.isArray(newField.options)) {
        newField.options = field.options.map((option) => ({
          ...option,
          label: t(option.label)
        }));
      }
      if (
        typeof newField.dataSource === 'object' &&
        Array.isArray(newField.dataSource.getOptions)
      ) {
        newField.dataSource.getOptions = field.dataSource.getOptions.map(
          (option) => ({
            ...option,
            label: t(option.label)
          })
        );
      }
      return newField;
    })
  }));
};
