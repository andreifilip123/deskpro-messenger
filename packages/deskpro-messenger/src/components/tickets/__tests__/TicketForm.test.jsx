import React from 'react';
import { render, fireEvent, wait } from '../../../utils/tests';
import TicketForm from '../TicketForm';

// commented out since we're using another form, need to update this test
// const ticketFormConfig = [
//   {
//     fields: [
//       {
//         name: 'subject',
//         label: 'Subject',
//         type: 'text',
//         validation: ['required'],
//         placeholder: 'Subject'
//       },
//       {
//         name: 'name',
//         label: 'Full Name',
//         type: 'text',
//         validation: ['required'],
//         placeholder: 'John Doe'
//       },
//       {
//         name: 'email',
//         label: 'Email',
//         type: 'text',
//         validation: ['required'],
//         placeholder: 'john.doe@company.com'
//       },
//       {
//         name: 'message',
//         label: 'Message',
//         type: 'textarea',
//         validation: ['required'],
//         placeholder: 'Enter you message here...'
//       }
//     ]
//   }
// ];
//
// describe('<TicketForm />', () => {
//   it('should render correctly', () => {
//     const { asFragment } = render(
//       <TicketForm formConfig={ticketFormConfig} onSubmit={jest.fn()} />
//     );
//     expect(asFragment()).toMatchSnapshot();
//   });
//
//   it('should submit values according the form config fields', async () => {
//     const onSubmit = jest.fn();
//     const { getByPlaceholderText, getByText } = render(
//       <TicketForm formConfig={ticketFormConfig} onSubmit={onSubmit} />
//     );
//     const button = getByText('Save Ticket');
//     expect(button).toBeInTheDocument();
//
//     fireEvent.change(getByPlaceholderText('Subject'), {
//       target: { value: 'Ticket from John Doe' }
//     });
//     fireEvent.change(getByPlaceholderText('John Doe'), {
//       target: { value: 'John Doe' }
//     });
//     fireEvent.change(getByPlaceholderText('john.doe@company.com'), {
//       target: { value: 'john.doe@deskpro.com' }
//     });
//     fireEvent.change(getByPlaceholderText('Enter you message here...'), {
//       target: { value: 'Lorem ipsum dolar sit amet.' }
//     });
//     fireEvent.click(button.parentElement);
//     await wait(() => {}, { timeout: 1000 });
//     expect(onSubmit).toBeCalledWith({
//       subject: 'Ticket from John Doe',
//       name: 'John Doe',
//       email: 'john.doe@deskpro.com',
//       message: 'Lorem ipsum dolar sit amet.'
//     });
//   });
// });
