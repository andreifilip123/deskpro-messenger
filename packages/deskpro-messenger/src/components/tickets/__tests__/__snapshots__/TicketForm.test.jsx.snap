// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<TicketForm /> should render correctly 1`] = `
<DocumentFragment>
  <form
    class="dp-pc_form show-hover"
  >
    <span>
      <div
        class="dp-pc_field dp-pc_empty"
      >
        <input
          class="dp-pc_input"
          description=""
          label="Subject"
          name="subject"
          placeholder="Subject"
          type="text"
          validation="required"
          value=""
          values="[object Object]"
        />
        <span
          class="indicator"
        />
        <label
          class="dp-pc_label"
          for="field_1"
        >
          Subject
        </label>
        <div
          class="divider"
        />
      </div>
    </span>
    <span>
      <div
        class="dp-pc_field dp-pc_empty"
      >
        <input
          class="dp-pc_input"
          description=""
          label="Full Name"
          name="name"
          placeholder="John Doe"
          type="text"
          validation="required"
          value=""
          values="[object Object]"
        />
        <span
          class="indicator"
        />
        <label
          class="dp-pc_label"
          for="field_2"
        >
          Full Name
        </label>
        <div
          class="divider"
        />
      </div>
    </span>
    <span>
      <div
        class="dp-pc_field dp-pc_empty"
      >
        <input
          class="dp-pc_input"
          description=""
          label="Email"
          name="email"
          placeholder="john.doe@company.com"
          type="text"
          validation="required"
          value=""
          values="[object Object]"
        />
        <span
          class="indicator"
        />
        <label
          class="dp-pc_label"
          for="field_3"
        >
          Email
        </label>
        <div
          class="divider"
        />
      </div>
    </span>
    <span>
      <div
        class="dp-pc_field dp-pc_empty"
      >
        <textarea
          class="dp-pc_input"
          component="textarea"
          description=""
          label="Message"
          name="message"
          placeholder="Enter you message here..."
          style="height: 0px;"
          type="textarea"
          validation="required"
          values="[object Object]"
        />
        <span
          class="indicator"
        />
        <label
          class="dp-pc_label"
          for="field_4"
        >
          Message
        </label>
        <div
          class="divider"
        />
      </div>
    </span>
    <button
      class="dpmsg-Button Button-FullWidth Button--medium Button--primary"
      type="submit"
    >
      <span>
        Save Ticket
      </span>
    </button>
  </form>
</DocumentFragment>
`;
